records/chains.go

```diff
@@ -0,0 +1,80 @@
+package records
+
+import (
+	"github.com/mesosphere/mesos-dns/records/labels"
+)
+
+// chain is a generation func that consumes record-like strings and does
+// something with them
+type chain func(...string)
+
+const (
+	protocolNone = "" // for readability
+	domainNone   = "" // for readability
+)
+
+// withProtocol appends `._{protocol}.{framework}` to records. if protocol is "" then
+// the protocols "tcp" and "udp" are assumed.
+func withProtocol(protocol, framework string, spec labels.Func, gen chain) chain {
+	return func(records ...string) {
+		protocol = spec(protocol)
+		if protocol != protocolNone {
+			for i := range records {
+				records[i] += "._" + protocol + "." + framework
+			}
+		} else {
+			records = append(records, records...)
```

Comment:
this *can* be more efficient if the underlying capacity of the slice is
large enough to hold twice the record count. in this particular instance it
depends how golang is allocating the initial slice for varargs parameter.
experiments on playground suggest that it allocates a slice w/ a capacity
equivalent to the # of variadic params.

even so, I'm loath to changing this line: it's concise and idiomatic.


On Thu, Mar 3, 2016 at 1:00 PM, Sargun Dhillon <notifications@github.com>
wrote:

> In records/chains.go
> <https://github.com/mesosphere/mesos-dns/pull/386#discussion_r54919995>:
>
> > +const (
> > +	protocolNone = "" // for readability
> > +	domainNone   = "" // for readability
> > +)
> > +
> > +// withProtocol appends `._{protocol}.{framework}` to records. if protocol is "" then
> > +// the protocols "tcp" and "udp" are assumed.
> > +func withProtocol(protocol, framework string, spec labels.Func, gen chain) chain {
> > +	return func(records ...string) {
> > +		protocol = spec(protocol)
> > +		if protocol != protocolNone {
> > +			for i := range records {
> > +				records[i] += "._" + protocol + "." + framework
> > +			}
> > +		} else {
> > +			records = append(records, records...)
>
> Is there any reason to reuse records, and do this in place? Does Golang
> actually care that much or optimize that much do this in place? Can we just
> create another slice, and populate that? I think that would make this code
> much easier to read.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mesosphere/mesos-dns/pull/386/files#r54919995>.
>